adding attribute examples

natsuki-engr opened the pull request:
- [x] faq
- [x] common techniques
- [x] cookbook

DerManoMann reviewed (changes requested):
Good stuff! Could you just please rename the files to something more meaningful? For example `server_xx.php`

natsuki-engr commented:
@DerManoMann
Thanks for checking my PR!
In the faq, the annotation sample is shown as an example of "dummy class".
So I chose the filename "dummy_class_xx.php" - was that not appropriate?

> The simplest solution to avoid this issue is to add a 'dummy' class to the docblock

Please let me know if my understanding is incorrect. 🙏

DerManoMann commented:
My bad - yes, this is fine.

Kiina commented on docs/snippets/guide/faq/dummy_class_at.php:
```diff
@@ -0,0 +1,18 @@
+<?php
+
+use OpenApi\Attributes as OAT;
```
The other docs all use as OA, while here it's as OAT. If you imported `OpenApi\Attributes\Tag as OAT;` I would get where the T is coming from, but here it might be wrong?

DerManoMann replied:
I do that actually everywhere in the codebase to distinguish annotations from attributes.
Very useful for slow migration where you might use both at the same time.

Kiina replied:
Well nvm then, seems like halfway through the user guide it sometimes switches to OAT. Cookbook even has both right after another:

https://zircote.github.io/swagger-php/guide/cookbook.html#mostly-virtual-models

DerManoMann replied:
Hmm, people seem to use both, but it would be good to settle on a consistent pattern in the docs, I think.

natsuki-engr replied:
should make another PR converting `OpenApi\Attributes as OA;` to `OpenApi\Attributes as OAT;` or fix my PR?

DerManoMann replied:
Ah, fix the PR. You can add more if you want to, too.

DerManoMann commented:
I think it would be good to add a script to validate that the snippets **do** evaluate to the same spec... I can look into that at some point.

DerManoMann commented:
One think I noticed is that the attribute examples use double quotes for string. Code style here is to use single quotes unless there is actual need to use double. Might be a pain, biut probbly good to get in. I think IDE can help with that.

OskarStark commented on docs/snippets/guide/cookbook/file_upload_with_headers_an.php:
```diff
@@ -0,0 +1,39 @@
+     *       )
+     *     )
+     *   ),
+     *   @OA\Response(
+     *     response=200, description="Success",
+     *     @OA\Schema(type="string")
+     *   ),
```
This is not in the attributes example file, isn't it?

natsuki-engr replied:
Thank you reviewing!
Exactly, and I realized that `@OA\Schema(type="string")` in annotation is probably unnecessary. I generated spec from `file_upload_with_headers_an.php` and 200 response has no content except 'description'.
![image](https://github.com/user-attachments/assets/800fa9b8-e1d4-4cb1-9bf9-312e2cc30752)

If my understanding is correct, should we remove it from the annotation in a separate PR?

DerManoMann commented on docs/guide/cookbook.md:
````diff
@@ -201,29 +201,19 @@ The `OA\Xml` annotation may be used to set the XML root element for a given `@OA
- */
-```
+
+<codeblock id="set-xml-root-name">
````
```suggestion
<codeblock id="upload-multipart-form-data">
```

natsuki-engr commented on docs/guide/cookbook.md:
Rather than splitting the example into PHP file as in other example, the sample here is leave in md file because it is short.

natsuki-engr commented:
@DerManoMann
CI test `tests/DocSnippetsTest::testSnippets` fails on some snippets because of mismatch of `operationId`.
I guess the snippets or test should be fixed.

For example, an error on file_upload_with_headers_at.php:
```
1) OpenApi\Tests\DocSnippetsTest::testSnippets with data set "file_upload_with_headers-3.0.0" (array('/workspaces/swagger-php/tests...an.php', '/workspaces/swagger-php/tests...at.php'), '3.0.0')
Snippet: $/workspaces/swagger-php/tests/../docs/snippets/guide/cookbook/file_upload_with_headers_at.php > paths > /v1/media/upload > post > operationId
Failed asserting that two strings are equal.
--- Expected
+++ Actual
@@ @@
-'6e2b268bb931464252c917c84b82cba5'
+'02d11fc2bbdec221ec8917f8a596bddd'

/workspaces/swagger-php/tests/OpenApiTestCase.php:175
/workspaces/swagger-php/tests/OpenApiTestCase.php:165
/workspaces/swagger-php/tests/OpenApiTestCase.php:165
/workspaces/swagger-php/tests/OpenApiTestCase.php:165
/workspaces/swagger-php/tests/OpenApiTestCase.php:165
/workspaces/swagger-php/tests/DocSnippetsTest.php:57
```

The `operationId` mismatch is caused by a difference in the namecepace of the copies created by `DocSnippetsTest::testSnippets`. Possible solutions are:
1. Modify `assertSpecEquals` to skip the `operationId` check
2. Explicitly specify the `operationId` value within the snippet (as in `minimal_api_*.php`)

However, neigher apprach seems particulary good. Do you have any better ideas?

DerManoMann commented:
@natsuki-engr I did push a fix to `master` that disables operationId generation - that should help. Sorry, no better idea 🤷 

I guess you can still put in explicit ids where it makes sense, but for the purpose of this test they are not that relevant IMO.

Nice work, btw.

DerManoMann commented on docs/snippets/guide/cookbook/external_documentation_an.php:
```diff
@@ -0,0 +1,13 @@
+<?php
+
```
Missing
```
use OpenApi\Annotations as OA;
```

DerManoMann commented on docs/snippets/guide/cookbook/file_upload_with_headers_an.php:
```diff
@@ -0,0 +1,39 @@
+<?php
+
```
Missing
```
use OpenApi\Annotations as OA;
```

DerManoMann commented on docs/snippets/guide/cookbook/nested_objects_an.php:
```diff
@@ -0,0 +1,36 @@
+<?php
+
```
Missing
```
use OpenApi\Annotations as OA;
```

DerManoMann commented on docs/snippets/guide/cookbook/oneof_example_an.php:
```diff
@@ -0,0 +1,23 @@
+<?php
+
```
Missing
```
use OpenApi\Annotations as OA;
```

DerManoMann commented on docs/snippets/guide/cookbook/properties_with_union_types_an.php:
```diff
@@ -0,0 +1,27 @@
+<?php
+
```
Missing
```
use OpenApi\Annotations as OA;
```

DerManoMann commented on docs/snippets/guide/cookbook/response_examples_an.php:
```diff
@@ -0,0 +1,29 @@
+<?php
+
+class Controller
```
Missing
```
use OpenApi\Annotations as OA;
```

DerManoMann commented on docs/snippets/guide/cookbook/reusing_response_an.php:
```diff
@@ -0,0 +1,29 @@
+<?php
+
```
Missing
```
use OpenApi\Annotations as OA;
```

DerManoMann commented on docs/snippets/guide/cookbook/secure_endpoint_an.php:
```diff
@@ -0,0 +1,16 @@
+<?php
+
+class Controller
```
Missing
```
use OpenApi\Annotations as OA;
```

DerManoMann commented on docs/snippets/guide/cookbook/security_schema_tips_an.php:
```diff
@@ -0,0 +1,18 @@
+<?php
+
```
Missing
```
use OpenApi\Annotations as OA;
```

DerManoMann commented on docs/snippets/guide/cookbook/security_schemas_an.php:
```diff
@@ -0,0 +1,26 @@
+<?php
+
```
Missing
```
use OpenApi\Annotations as OA;
```

DerManoMann commented on docs/snippets/guide/cookbook/set_xml_root_name_an.php:
```diff
@@ -0,0 +1,22 @@
+<?php
+
```
Missing
```
use OpenApi\Annotations as OA;
```

DerManoMann commented on docs/snippets/guide/cookbook/uploading_multipart_formdata_an.php:
```diff
@@ -0,0 +1,31 @@
+<?php
+
+
```
Missing
```
use OpenApi\Annotations as OA;
```

DerManoMann commented on docs/snippets/guide/cookbook/x_tag_groups_an.php:
```diff
@@ -0,0 +1,14 @@
+<?php
+
```
Missing
```
use OpenApi\Annotations as OA;
```

DerManoMann commented on docs/snippets/guide/faq/dummy_class_an.php:
```diff
@@ -0,0 +1,21 @@
+<?php
+
+/**
```
Missing
```
use OpenApi\Annotations as OA;
```

DerManoMann reviewed (changes requested):
A few places where `use OpenApi\Annotations as OA;` is missing - would be great to get those added.

Another thing is that we now should add the `docs/snippet` folder to CS - for that you'd need to update `.php-cs-fixer.dist.php` to add the folder; something like this:
```php
// ...
$finder = PhpCsFixer\Finder::create()
    ->path('src')->name('*.php')
    ->path('tools')->name('*.php')
    ->path('docs/examples')->name('*.php')
    ->path('docs/snippets')->name('*.php')
    ->path('tests')->name('*.php')

```

Other than that it looks great. Thanks for taking the time.

natsuki-engr commented:
@DerManoMann 
Thank you so much for fixing the test code. It seems to be working fine.

I apologize for the missing many use statements. I've added them, please check it again.
I also added CS check for `docs/snippets` directory, and the resulting formatting changes are included in commit 66d7e95.
Thanks again for your help!

DerManoMann commented:
No problem. Any help is appreciated and improving the docs is always valuable.

DerManoMann commented:
@natsuki-engr Deployed and looking real nice!

As a side note I noticed some examples have corresponding YAML - it wonder if adding a new tab for the generated spec would work ...
